### ISSUE TITLE
Could you add more servers in FedAvg for faster training speed?

### ISSUE THREAD

**Issue by wizard1203** [FedML-AI/FedML#59]
Could you add more servers in FedAvg for faster training speed?
As the BytePS does.

**Comment by chaoyanghe** [comment]
BytePS is for data center-based distributed training, while FedML (e.g., FedAvg) is edge-based distributed training. The particular assumptions of FL include:
1. heterogeneous data distribution cross devices (non-I.I.D.)
2. resource constrained edge devices (memory, computational, and communication)
3. label deficiency (harder to label the data points because of privacy)
4. concern the security and privacy

So what do you mean "adding more servers in FedAvg"?

**Comment by wizard1203** [comment]
> BytePS is for data center-based distributed training, while FedML (e.g., FedAvg) is edge-based distributed training. The particular assumptions of FL include:
> 
> 1. heterogeneous data distribution cross devices (non-I.I.D.)
> 2. resource constrained edge devices (memory, computational, and communication)
> 3. label deficiency (harder to label the data points because of privacy)
> 4. concern the security and privacy
> 
> So what do you mean "adding more servers in FedAvg"?

I mean adding more parameter servers to improve the communication efficiency. Maybe this can be used suitably only in cluster environment but not true Federated Learning environment with resource constrained edge devices. However, it is still can accelerate the training when doing research.

**Comment by prosopher** [comment]
FedML supports multiple parameter servers for the communication efficiency via hierarchical FL and decentralized FL .
In hierarchical FL, there are group parameter servers that split the total client set into multiple client subsets.
In decentralized FL, each client acts as a parameter server.

Please refer to the following links for details.
https://github.com/FedML-AI/FedML/tree/master/fedml_experiments/standalone/decentralized
https://github.com/FedML-AI/FedML/tree/master/fedml_experiments/standalone/hierarchical_fl

**Comment by chaoyanghe** [comment]
@wizard1203 Thanks for your suggestion. As for acceleration, we are the only research-oriented framework that supports cross machine multiple GPU distributed training. To further accelerate, definitely we can use many techniques from traditional distributed training (very mature with much less research attention). I just elaborate a few here:
1. with AllReduce-based GPU-GPU communication using InfiniBand. However, this is not a real FL setting. As you said, it is only useful to evaluate algorithms or modeling which are not insensitive to the training speed.
2. Hybrid parallelism (model + data parallelism) + pipeline
3. Bucking BP (as introduced in PyTorch VLDB 2020)
4. low-bit (half precision)
5. pruning
....

As for now, since we already support distributed training. You can config your own training topology at the main_xxx.py. 

I have to point out that "adding more parameter servers to improve the communication efficiency" is a bit confusing conceptually. We cannot say more computation resource improves the communication efficiency. The relationship between computation and communication is normally a trade-off. When using more parallel computation does not mean we can speedup the training since the communication cross machines may dominate the training time. But I agree your idea that using traditional techniques in distributed computing to accelerate FL research. Thanks.

**Comment by wizard1203** [comment]
> FedML supports multiple parameter servers for the communication efficiency via hierarchical FL and decentralized FL .
> In hierarchical FL, there are group parameter servers that split the total client set into multiple client subsets.
> In decentralized FL, each client acts as a parameter server.
> 
> Please refer to the following links for details.
> https://github.com/FedML-AI/FedML/tree/master/fedml_experiments/standalone/hierarchical_fl
> https://github.com/FedML-AI/FedML/tree/master/fedml_experiments/standalone/decentralized

@prosopher Thanks for this, I will carefully read them.

**Comment by chaoyanghe** [comment]
> FedML supports multiple parameter servers for the communication efficiency via hierarchical FL and decentralized FL .
> In hierarchical FL, there are group parameter servers that split the total client set into multiple client subsets.
> In decentralized FL, each client acts as a parameter server.
> 
> Please refer to the following links for details.
> https://github.com/FedML-AI/FedML/tree/master/fedml_experiments/standalone/hierarchical_fl
> https://github.com/FedML-AI/FedML/tree/master/fedml_experiments/standalone/decentralized

Thanks. But I guess he was discussing the distributed computing setting, not the standalone version.

**Comment by wizard1203** [comment]
@chaoyanghe Thanks for your detailed explanation. Maybe I can try to complete it by myself, and when I finish it I would like to push it to your master branch.

**Comment by chaoyanghe** [comment]
> @chaoyanghe Thanks for your detailed explanation. Maybe I can try to complete it by myself, and when I finish it I would like to push it to your master branch.

Thanks. Looking forward to your contribution.

**Comment by chaoyanghe** [comment]
@wizard1203 Do you mean modifying based on this code?
https://github.com/FedML-AI/FedML/tree/master/fedml_experiments/distributed/fedavg

**Comment by wizard1203** [comment]
> @wizard1203 Do you mean modifying based on this code?
> https://github.com/FedML-AI/FedML/tree/master/fedml_experiments/distributed/fedavg

@chaoyanghe No, maybe it needs to base on those codes on fedml_core. Whatever, I may try to do it many days later. In fact, I have some other algorithms that I want to implement more urgently than this.
